### PR TITLE
Support bit shifting `SymInt`s

### LINKED ISSUE #104228: Bit shifting `SymInt`s causes graph breaks
### 🐛 Describe the bug

Dynamo currently graph breaks on `int << SymInt`, `SymInt << int` and `SymInt << SymInt` (the same holds for `>>`):

```python
import torch

@torch.compile(fullgraph=True, dynamic=True, backend="eager")
def f(x):
    return x << 0

f(2)
```

It would be nice to handle these cases.

### Error logs

```
Traceback (most recent call last):
  File "bug.py", line 7, in <module>
    f(2)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 295, in _fn
    return fn(*args, **kwargs)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 448, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 127, in _fn
    return fn(*args, **kwargs)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 360, in _convert_frame_assert
    return _compile(
  File "/.../lib/python3.8/site-packages/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 430, in _compile
    out_code = transform_code_object(code, transform)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 415, in transform
    tracer.run()
  File "/.../lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 2027, in run
    super().run()
  File "/.../lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 708, in run
    and self.step()
  File "/.../lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 668, in step
    getattr(self, inst.opname)(inst)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 161, in impl
    self.push(fn_var.call_function(self, self.popn(nargs), {}))
  File "/.../lib/python3.8/site-packages/torch/_dynamo/variables/builtin.py", line 618, in call_function
    return super().call_function(tx, args, kwargs)
  File "/.../lib/python3.8/site-packages/torch/_dynamo/variables/base.py", line 237, in call_function
    unimplemented(f"call_function {self} {args} {kwargs}")
  File "/.../lib/python3.8/site-packages/torch/_dynamo/exc.py", line 140, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_function BuiltinVariable(lshift) [SymNodeVariable(), ConstantVariable(int)] {}

from user code:
   File "bug.py", line 5, in f
    return x << 0

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

### Minified repro

_No response_

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230626+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, May 26 2023, 14:05:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.29
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 530.30.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          32
On-line CPU(s) list:             0-31
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Silver 4314 CPU @ 2.40GHz
Stepping:                        6
CPU MHz:                         2899.687
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4800.00
Virtualization:                  VT-x
L1d cache:                       768 KiB
L1i cache:                       512 KiB
L2 cache:                        20 MiB
L3 cache:                        24 MiB
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] torch==2.1.0.dev20230626+cpu
[conda] Could not collect
```

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by ar0ck**
Fixes #104228.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -279,7 +279,7 @@ def parallel_or(*args):
     Evaluate the logical OR of several arguments, avoiding guarding on
     unbacked SymInts if another argument is definitely True.
     """
-    if any(definitely_true(args) for a in args):
+    if any(definitely_true(a) for a in args):
```
Comment (ar0ck) [inline torch/fx/experimental/symbolic_shapes.py:282]: @ezyang - sneaking this fix in here, assuming this was just a typo in unused code and not something galaxy-brain?

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:282]: oops nice catch

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
needs rebase

**Comment by ar0ck** [comment]
> needs rebase

Done - wasn't sure whether `LShift` + `RShift` belong in `symbolic_shapes.py` or `_sympy/functions.py`, went with `_sympy/functions.py`.

**Comment by ezyang** [comment]
Latter is correct

**Comment by ar0ck** [comment]
Test failed due to `vision_maskrcnn` going from `56` to `51` graphs breaks. Didn't look into it, but sounds conceivable so I just updated the expected result.

**Pull request closed without merging**